lib/TopSitesFeed.jsm

```diff
@@ -188,17 +192,30 @@ this.TopSitesFeed = class TopSitesFeed {
       return copy;
     }));
 
-    // Remove any duplicates from frecent and default sites
-    const [, dedupedFrecent, dedupedDefaults] = this.dedupe.group(
-      pinned, frecent, notBlockedDefaultSites);
-    const dedupedUnpinned = [...dedupedFrecent, ...dedupedDefaults];
-
-    // Remove adult sites if we need to
-    const checkedAdult = this.store.getState().Prefs.values.filterAdult ?
-      filterAdult(dedupedUnpinned) : dedupedUnpinned;
+    const dedupedUnpinned = this.dedupeTopSites({
+      // Convert any frecent site that matches SEARCH_HOST_FILTERS to a search topsite
+      // without pinning or changing position
+      frecent: frecent.map(this.topSiteToSearchTopSite),
+      pinned,
+      defaultSites: notBlockedDefaultSites
+    });
 
     // Insert the original pinned sites into the deduped frecent and defaults
-    const withPinned = insertPinned(checkedAdult, pinned).slice(0, numItems);
+    // and check if any match default search topsite rules
+    for (const site of insertPinned(dedupedUnpinned, pinned).slice(0, numItems)) {
+      // A default site that is also a search topsite should be pinned to
+      // the first position
+      if (site.searchTopSite && site.isDefault && !site.isPinned) {
+        await this._insertPin(site, 0);
+        // Refresh the cache because we just modified pinned sites
+        this.pinnedCache.expire();
+        pinned = await this.pinnedCache.request();
```

Comment:
Does this result in multiple Places sql queries? If so, does it work if we just expire the cache and re-request once after the for loop instead?